### PR TITLE
[ShaderGraph] [2022.1] Making noise nodes platform deterministic

### PR DESCRIPTION
**Pull request by cdxntchou**
Original PR here: https://github.com/Unity-Technologies/Graphics/pull/5452
Switched branches so I could rebase off of a passing master build.

### Purpose of this PR
Fix for bug:  https://fogbugz.unity3d.com/f/cases/1156544/

Console PR: https://github.cds.internal.unity3d.com/unity/ScriptableRenderPipelinePrivate/pull/234

Introduces a set of deterministic integer hashes, with varying input and output dimensions.
These hashes are integer based, so they produce the same result on ALL platforms.

Converted noise nodes to have enum switches to choose between legacy hash behavior, and the new deterministic hashes.
Existing noise nodes will use legacy behavior for backwards compatibility, newly created nodes choose deterministic by default.

New deterministic hashes also have a more random appearance:  (legacy hash looks bad on nvidia cards)
image

Doc ticket added here:  https://jira.unity3d.com/browse/GSG-548

---
### Testing status

Added new test cases to the Procedural test scene, to test both legacy and deterministic versions of the noise nodes.
(noise node tests were previously disabled... enabled them)

Validated existing noise nodes upgrade with identical results, using the legacy hashes.
image

Tested newly created noise nodes, default to deterministic hashes:
image

Yamato:

ShaderGraph:  🟢 
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252Fnoise2/.yamato%252Fall-shadergraph.yml%2523Nightly_ShaderGraph_2021.2/8494066/job/pipeline
Master:   35e38b19cc9212a9026b7f12d92e9fa79e677cfe
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fall-shadergraph.yml%2523Nightly_ShaderGraph_2021.2/8364525/job/pipeline

ShaderGraph Trunk: 🟢
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252Fnoise2/.yamato%252Fall-shadergraph.yml%2523PR_ShaderGraph_trunk/8664161/job/pipeline
Master: 9f86ed365a8a2bc00e1aeb94c39c0bf31fdb1114
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fall-shadergraph.yml%2523PR_ShaderGraph_trunk/8643605/job/pipeline

Console Tests: 🟡 - same failures as master
https://github.cds.internal.unity3d.com/unity/ScriptableRenderPipelinePrivate/pull/234

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Review by bencloward — APPROVED** [review]
This is going to improve out noise a lot.  Thanks, Chris!

**Review by sarah-welton — APPROVED** [review]
Docs changes are in as requested. Seems like a great change, and happy to see it!

File: com.unity.shadergraph/Documentation~/Gradient-Noise-Node.md [hunk 1]
```diff
@@ -12,6 +14,13 @@ Generates a gradient, or [Perlin](https://en.wikipedia.org/wiki/Perlin_noise), n
 | Scale      | Input | Float    | None | Noise scale |
 | Out | Output      |    Float    | None | Output value |
 
+## Controls
+
+| Name        | Type           | Options  | Description |
+|:------------ |:-------------|:-----|:---|
+| Hash Type      | Dropdown | Deterministic, LegacyMod | Selects the hash function used to generate random numbers for noise generation. |
```
Comment (jessebarker) [inline com.unity.shadergraph/Documentation~/Gradient-Noise-Node.md:21]: Should this say "LegacyMode" (rather than LegacyMod)?

**Review by jessebarker — APPROVED** [review]
LGTM

File: com.unity.shadergraph/Documentation~/Gradient-Noise-Node.md (hunk 1 above)
Comment (cdxntchou) [reply com.unity.shadergraph/Documentation~/Gradient-Noise-Node.md:21]: it is actually a modulo operation.

**Review by sarah-welton — APPROVED** [review]
Forgot I already reviewed because I still had this open. Ignore me

**Pull request merged**